File: src/jvmTest/java/org/snakeyaml/engine/schema/FailsafeTest.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.snakeyaml.engine.schema;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;
import it.krzeminski.snakeyaml.engine.kmp.api.Dump;
import it.krzeminski.snakeyaml.engine.kmp.api.DumpSettings;
import it.krzeminski.snakeyaml.engine.kmp.api.Load;
import it.krzeminski.snakeyaml.engine.kmp.api.LoadSettings;
import it.krzeminski.snakeyaml.engine.kmp.schema.FailsafeSchema;

@org.junit.jupiter.api.Tag("fast")
public class FailsafeTest {

  Load loader = new Load(LoadSettings.builder().setSchema(new FailsafeSchema()).build());

  @Test
  void parseString() {
    assertEquals("true", loader.loadOne("true"));
    assertEquals("false", loader.loadOne("false"));
    assertEquals("null", loader.loadOne("null"));
    assertEquals("1", loader.loadOne("1"));
    assertEquals("0001", loader.loadOne("0001"));
    assertEquals("3.000", loader.loadOne("3.000"));
  }

  @Test
  void dumpString() {
    Dump dumper = new Dump(DumpSettings.builder().setSchema(new FailsafeSchema()).build());
    assertEquals("!!bool 'true'\n", dumper.dumpToString(Boolean.TRUE));
    assertEquals("!!bool 'false'\n", dumper.dumpToString(Boolean.FALSE));
    assertEquals("!!null 'null'\n", dumper.dumpToString(null));
    assertEquals("!!int '25'\n", dumper.dumpToString(25));
    assertEquals("!!int '17'\n", dumper.dumpToString(Integer.valueOf(17)));
    assertEquals("!!float '17.4'\n", dumper.dumpToString(Double.valueOf(17.4)));
  }
}

File: src/commonTest/kotlin/it/krzeminski/snakeyaml/engine/kmp/schema/FailsafeTest.kt
package it.krzeminski.snakeyaml.engine.kmp.schema

import io.kotest.core.spec.style.FunSpec
import io.kotest.data.forAll
import io.kotest.matchers.shouldBe
import io.kotest.data.headers
import io.kotest.data.row
import io.kotest.data.table
import it.krzeminski.snakeyaml.engine.kmp.api.Dump
import it.krzeminski.snakeyaml.engine.kmp.api.DumpSettings
import it.krzeminski.snakeyaml.engine.kmp.api.Load
import it.krzeminski.snakeyaml.engine.kmp.api.LoadSettings

class FailsafeTest : FunSpec({
    val loader = Load(LoadSettings(schema = FailsafeSchema()))

    test("parse string") {
        forAll(
            table(
                headers("input", "expected"),
                row("true", "true"),
                row("false", "false"),
                row("null", "null"),
                row("1", "1"),
                row("0001", "0001"),
                row("3.000", "3.000")
            )
        ) { input, expected ->
            loader.loadOne(input) shouldBe expected
        }
    }

    test("dump string") {
        val dumper = Dump(DumpSettings(schema = FailsafeSchema()))
        forAll(
            table(
                headers("input", "expected"),
                row(true, "!!bool 'true'\n"),
                row(false, "!!bool 'false'\n"),
                row(null, "!!null 'null'\n"),
                row(12.toByte(), "!!int '12'\n"),
                row(45.toShort(), "!!int '45'\n"),
                row(25, "!!int '25'\n"),
                row(17, "!!int '17'\n"),
                row(34L, "!!int '34'\n"),
                row(17.4, "!!float '17.4'\n"),
                row(23.5, "!!float '23.5'\n"),
            )
        ) { input, expected ->
            dumper.dumpToString(input) shouldBe expected
        }
    }
})
